packages/jupyter-ai/jupyter_ai/personas/persona_manager.py

```diff
@@ -0,0 +1,174 @@
+from logging import Logger
+from time import time_ns
+from typing import TYPE_CHECKING, ClassVar, Optional
+
+from importlib_metadata import entry_points
+from jupyterlab_chat.models import Message
+from jupyterlab_chat.ychat import YChat
+
+from ..config_manager import ConfigManager
+from .base_persona import BasePersona
+
+if TYPE_CHECKING:
+    from asyncio import AbstractEventLoop
+
+# EPG := entry point group
+EPG_NAME = "jupyter_ai.personas"
+
+
+class PersonaManager:
+    """
+    Class that manages all personas for a single chat.
+    """
+
+    # instance attrs
+    ychat: YChat
+    config_manager: ConfigManager
+    event_loop: "AbstractEventLoop"
+    log: Logger
+    _personas: dict[str, BasePersona]
+
+    # class attrs
+    _persona_classes: ClassVar[Optional[list[type[BasePersona]]]] = None
+
+    def __init__(
+        self,
+        ychat: YChat,
+        config_manager: ConfigManager,
+        event_loop: "AbstractEventLoop",
+        log: Logger,
+    ):
+        self.ychat = ychat
+        self.config_manager = config_manager
+        self.event_loop = event_loop
+        self.log = log
+
+        if not isinstance(PersonaManager._persona_classes, list):
+            self._init_persona_classes()
+            assert isinstance(PersonaManager._persona_classes, list)
+
+        self._personas = self._init_personas()
+
+    def _init_persona_classes(self) -> None:
+        """
+        Initializes the list of persona *classes* by retrieving the
+        `jupyter-ai.personas` entry points group.
+
+        This list is cached in the `PersonaManager._persona_classes` class
+        attribute, i.e. this method should only run once in the extension
+        lifecycle.
+        """
+        if PersonaManager._persona_classes:
+            return
+
+        persona_eps = entry_points().select(group=EPG_NAME)
+        self.log.info(f"Found {len(persona_eps)} entry points under '{EPG_NAME}'.")
+        self.log.info("PENDING: Loading AI persona classes from entry points...")
+        start_time_ns = time_ns()
+        persona_classes: list[type[BasePersona]] = []
+
+        for persona_ep in persona_eps:
+            try:
+                persona_class = persona_ep.load()
+                assert issubclass(persona_class, BasePersona)
+                persona_classes.append(persona_class)
+                class_module, class_name = persona_ep.value.split(":")
+                self.log.info(
+                    f"  - Loaded AI persona class '{class_name}' from '{class_module}' using entry point '{persona_ep.name}'."
+                )
+            except Exception as e:
+                self.log.error(
+                    f"  - Unable to load AI persona from entry point `{persona_ep.name}` due to an exception printed below."
+                )
+                self.log.exception(e)
+                continue
```

Comment:
swallowing all the exceptions? Perhaps that's okay, but consider adding inline comment to explain why this is better than failing outright.